Reinstate the KDF error macros

levitte opened the pull request:
include/openssl/kdferr.h was removed because it got empty.  However,
that will break the build of any application that includes it
directly, so we add it back as a placeholder.

Fixes #10340

kroeckx commented:
I expect all the old defines to be in there. That is, the KDF_F_* defined to 0, the KDF_R_* to some distinct value >= 100. There also was an ERR_load_KDF_strings().

kroeckx reviewed (changes requested):
All old defines/functions should be in it.

levitte commented:
Ah, so what you're asking for is a revert of fe6ec26b204a056aee2a24b79df09a45b2308603, basically

levitte committed:
Reinstate the KDF error macros

For minimum breakage with existing applications that might use them.

This reverts commit fe6ec26b204a056aee2a24b79df09a45b2308603 and
37ed62107112d95f7b7c9bf75602a6ac40883a89.

Fixes #10340

kroeckx commented:
# Number of lines in @nm_lines before massaging: 4935
# Number of lines in @def_lines before massaging: 4795
# Number of lines in @nm_lines after massaging: 4790
# Number of lines in @def_lines after massaging: 4791
# The following symbols are missing in libcrypto.so:
#   ERR_load_KDF_strings

kroeckx commented on include/openssl/kdferr.h:
```diff
@@ -0,0 +1,110 @@
+# ifdef  __cplusplus
+extern "C"
+# endif
+int ERR_load_KDF_strings(void);
```
I guess you want to mark this as deprecated

levitte replied:
Good idea

kroeckx commented:
and run make update

levitte commented:
Ah, yeah, forgot that.

levitte committed:
util/perl/OpenSSL/ParseC.pm: Fix 'extern "C" DEPRECATEDIN_x_y(...)'

The parser expected an 'extern "C"' followed by a single declaration
to always end with a semicolon.  Then came along something like this:

    extern "C" DEPRECATEDIN_3_0(int ERR_load_KDF_strings(void))

This change adjusts the detector of 'extern "C"' to also take in
accound a declaration that ends with a parenthesis.

levitte committed:
Deprecate ERR_load_KDF_strings()

levitte committed:
Make sure KDF reason codes are conserved in their current state

Because KDF errors are deprecated and only conserved for backward
compatibilty, we must make sure that they remain untouched.  A simple
way to signal that is by modifying crypto/err/openssl.ec and replace
the main header file (include/openssl/kdf.h in this case) with 'NONE',
while retaining the error table file (crypto/kdf/kdf_err.c).

util/mkerr.pl is modified to silently ignore anything surrounding a
conserved lib when such a .ec line is found.

levitte commented:
I reverted a lot more than might be practically needed...  I could see reason to empty `crypto/kdf/kdf_err.c` of almost all content, just leaving `ERR_load_KDF_strings()` as a function that does nothing.

Still, it was an interesting exercise, to invent some kind of "conservation state" for this...  we may need to do the same with others, eventually.

kroeckx commented:
I don't see a reason to keep the strings in kdf_err.c

kroeckx commented:
```
crypto: function ERR_load_KDF_strings undocumented
# 1 in util/libcrypto.num are not documented
```

levitte commented:
Eh gods. Could we perhaps make a documentation exception on deprecated functions?
Meanwhile, I'll add it to the appropriate missing file.

levitte committed:
squash! Deprecate ERR_load_KDF_strings()

Not only deprecate, but also remove the reason strings and make
ERR_load_KDF_strings() do nothing.

levitte commented:
All done

kroeckx commented on crypto/err/err_all.c:
```diff
@@ -38,6 +38,7 @@
 #include <openssl/cmperr.h>
 #include <openssl/cterr.h>
 #include <openssl/asyncerr.h>
+#include <openssl/kdferr.h>
```
Why this include?

levitte replied:
Because I actually did a revert and didn't notice this :wink:

levitte replied:
Fixed

kroeckx commented on include/openssl/kdferr.h:
```diff
@@ -0,0 +1,116 @@
+/*
+ * KDF reason codes.
+ */
+# define KDF_R_BAD_ENCODING                               122
```
Mark as deprecated too?

levitte replied:
Good idea

levitte replied:
Done

levitte committed:
fixup! Reinstate the KDF error macros

levitte committed:
fixup! Deprecate ERR_load_KDF_strings()

kroeckx commented:
I think it looks good, but I'm not going to review (the perl part).

t8m reviewed (approved):
Looks good, including the perl.

levitte commented:
Merged.

7c6a0d909abda7c854b44831ef43ebdda49ed42e Reinstate the KDF error macros
dc5ce51983713dfbb8f3d0c8b198dee32b563ac6 util/perl/OpenSSL/ParseC.pm: Fix 'extern "C" DEPRECATEDIN_x_y(...)'
905b097fd586ed0279aee05b5961224212014ec9 Deprecate ERR_load_KDF_strings()
4d301427a96010468da2bb67bf1025fa8d886ab9 Make sure KDF reason codes are conserved in their current state
